Normalize priorities to not exceed total voting power

Liamsi opened the pull request:
Followup PR to #2966 
This should fix #2960

It basically introduces a rescaling/normalization of the priorities to make sure that the prios do not exceed 2*TotalVotingPower. Also, the averaging is done only once (at the end of an IncrementProposerPriority step instead of in each iteration).

@cwgoes and me wrote up a little proof/formalization why this works. It boils down to to showing:

For any sequence of proposer elections with no power changes and n blocks, the difference between n*power/totalPower and the number of proposed blocks (for each validator) is at most 1.

Work in progress as I still want to tex that proof and discuss it with the team first.
 
<!-- Thanks for filing a PR! Before hitting the button, please check the following items.-->

* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [x] Wrote tests
* [x] Updated CHANGELOG_PENDING.md

cwgoes commented:
I've written up a formalization of the fairness proof in Idris [here](https://github.com/cwgoes/tm-proposer-idris) - right now just for the two-validator case, but I believe @ValarDragon had a reduction from the n-validator case to the 2-validator case (which is also worth writing up).

Liamsi commented on types/validator_set.go:
```diff
@@ -78,44 +78,45 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
 	return mostest
 }
 
+func (vals *ValidatorSet) dividePrioritiesBy(divisor int64) {
```
note to myself: add a comment that this is not expected to be called with divisor == 0

cwgoes replied:
How about we panic? Dividing-by-zero is definitely justification for panicking. Will Golang panic anyways?

ebuchman replied:
Go will panic

ebuchman commented on types/validator_set.go:
```diff
@@ -78,44 +78,45 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
+	//  2*totalVotingPower/(maxPriority - minPriority)
+	diff := computeMaxMinPriorityDiff(vals)
+	threshold := 2 * vals.TotalVotingPower()
+	if diff > threshold && threshold > 0 {
```
Why would we ever have `threshold <= 0`

Liamsi replied:
As far as I remember, I've observed total voting power being 0 (at least in tests). Then, threshold would be 0, too.

ebuchman replied:
Ok. Shouldn't ever happen. Maybe there's some weird test where it does. I suppose it's fine for a sanity check but lets leave a note that it shouldn't happen and we ought to be able to remove that check

Liamsi replied:
E.g., the test `TestAveragingInIncrementProposerPriority` doesn't init voting powers. Only prios. as it only aims to test the priority averaging logic. But I remember there were other unrelated tests with voting power 0, too.

Liamsi commented:
> One thing I'm wondering about amidst all these renormalizing is how we're handling ties in priority - it looks like we break them based on an address comparison but could that compromise fairness if the same address always wins?

Hmm, looks you are right on first sight. But being elected proposer changes the priority and hence the next round won't consist of the same tie. Also, there are few tests that test that proposers with the same power alternate, e.g. `TestProposerPriorityProposerAlternates`: https://github.com/tendermint/tendermint/pull/3049/files#diff-a1153b96e079da8da43231ad48de57a0L334

ebuchman commented:
> But being elected proposer changes the priority and hence the next round won't consist of the same tie.

Yeh but I wonder if there's a situation where you can end up with the same priority again, before the other guy got to be proposer, and ultimately prevent them from ever being proposer, or at least prevent it for a long time.

milosevic commented on types/validator_set.go:
```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
 	}
 
-	const shiftEveryNthIter = 10
+	// cap the difference between priorities to be proportional to 2*totalPower by
```
As imbalance happens only upon validator set changes, I would suggest to move this code to `execution.go/updateValidators` and/or `validator_set.go/Add, validator_set.go/Remove`. Idea is to to averaging and normalisation only when val set is changed and then have "plain" proposer selection at the beginning of each round. As this might involve significant changes maybe it's better to address this in a separate issue/PR.

Liamsi replied:
Thanks for the clarification. That makes sense for the normalisation, not sure about the averaging/shifting but probably too.

Liamsi commented on types/validator_set.go:
```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
+		// div = Ceil((maxPriority - minPriority) / 2*totalVotingPower)
+		// threshold > 0 and diff > threshold guarantees (diff / threshold > 0):
+		div := int64(math.Ceil(float64(diff) / float64(threshold)))
+		vals.dividePrioritiesBy(div)
```
we could probably inline `dividePrioritiesBy` here as it's just a for-loop and we would save one casting to int and back to float64 inside of `dividePrioritiesBy`.

cwgoes commented on types/validator_set.go:
```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
-		vals.shiftByAvgProposerPriority(validatorsHeap)
+		proposer = vals.incrementProposerPriority()
 	}
+	vals.shiftByAvgProposerPriority()
```
Are we sure averaging only once per call is safe? If we don't average every `n` times, and `times` is really large (maybe when catching up syncing a node), it seems like we might run into the clipping logic and end up with a different final set of priorities (and possibly validators).

Liamsi replied:
Good point. Given the speed tendermint creates blocks, being offline for a few weeks (or months) may cause the clipping to kick in. Not sure how realistic that actually is but we should probably account for this.  Previously, we had every 10th iter. Maybe we should just make this number a bit larger then?

cwgoes commented on types/validator_set.go:
```diff
@@ -131,11 +139,34 @@ func (vals *ValidatorSet) computeAvgProposerPriority() int64 {
+func (vals *ValidatorSet) getValWitMostPriority() *Validator {
+	var res *Validator
+	for _, val := range vals.Validators {
+		res = res.CompareProposerPriority(val)
```
This is pretty unintuitive to read - do most objects in Tendermint have functions called `CompareXYZ` which return the higher-valued object by some property? Maybe I'm just not used to it. Otherwise I think directly comparing `val.ProposerPriority` or having a function `LargestValidatorByPriority` would be a bit more obvious.

Liamsi replied:
It is the same method as used in `GetProposer` (via `findProposer`). I agree that it is weird, though.

cwgoes commented on types/validator_set.go:
```diff
@@ -131,11 +139,34 @@ func (vals *ValidatorSet) computeAvgProposerPriority() int64 {
+	return diff
+}
+
+func (vals *ValidatorSet) getValWitMostPriority() *Validator {
```
```suggestion
func (vals *ValidatorSet) getValWithMostPriority() *Validator {
```

cwgoes commented on types/validator_set.go:
```diff
@@ -78,44 +78,52 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
-
 	// Decrement the validator with most ProposerPriority:
-	mostest := validatorsHeap.Peek().(*Validator)
+	mostest := vals.getValWitMostPriority()
```
```suggestion
	mostest := vals.getValWithMostPriority()
```

cwgoes commented on CHANGELOG_PENDING.md:
```diff
@@ -18,6 +18,8 @@ Special thanks to external contributors on this release:
   * [merkle] \#2713 Merkle trees now match the RFC 6962 specification
 
 * P2P Protocol
+  - [consensus] \#2960 normalize priorities to not exceed `2*TotalVotingPower` to mitigate unfair proposer selection 
```
Note that `2 * TotalVotingPower` doesn't give the strictest possible fairness bound - one validator might still get an extra block after some power changes - but I think the bound it gives is enough in practice.

ebuchman replied:
Why are we using 2* ? Doesn't the idris proof just use the 1* ?

cwgoes replied:
I think because we wanted to retain the `-9/8 * TotalPower` initial priority for newly added validators - which would be negated if we immediately renormalized to max-diff `TotalPower` afterwards.

ebuchman commented:
@Liamsi @cwgoes would appreciate a look at the PR I just merged in https://github.com/tendermint/tendermint/pull/3155

Some final points before this is truly complete:
- review #3155 
- avg every n times instead of just once per call to increment
- review/eliminate use of floats (!)

Liamsi commented:
> review/eliminate use of floats (!)

I removed the usage of float64 altogether. To compute the `div`,  I manually ceil (add 1 if `diff` isn't a multiple of threshold) and I removed the 2nd ceil in `dividePrioritiesBy`.

ebuchman commented:
Merging this. Thanks for the hard work everyone. Let's follow up in a new PR if need be.

melekes commented on types/validator_set.go:
```diff
@@ -78,44 +78,57 @@ func (vals *ValidatorSet) IncrementProposerPriority(times int) {
+	// removed if all tests would init. voting power appropriately;
+	// i.e. diffMax should always be > 0
+	if diffMax == 0 {
+		return
```
Is the purpose here to never panic? What if diffMax is 0, will we be able to find it out?

Liamsi replied:
It is probably appropriate to panic here. Some tests do not init. voting power accordingly though. We should fix those (as all those would panic). Outside of tests this (diffMax <= 0= should never happen.
